tsdb/index/postings_test.go

```diff
@@ -1282,3 +1282,37 @@ func BenchmarkListPostings(b *testing.B) {
 		})
 	}
 }
+
+func BenchmarkMemPostings_Add(b *testing.B) {
+	var series []struct {
+		id   storage.SeriesRef
+		lset labels.Labels
+	}
+	for i := 0; i < 1000; i++ {
+		series = append(series, struct {
+			id   storage.SeriesRef
+			lset labels.Labels
+		}{
+			id: storage.SeriesRef(i),
+			lset: labels.FromStrings(
+				"job", fmt.Sprintf("job%d", i%100),
+				"instance", fmt.Sprintf("instance%d", i),
+				"cluster", "dev",
+				"label1", fmt.Sprintf("value%d", i%5),
+				"label2", fmt.Sprintf("value%d", i%5),
+				"label3", fmt.Sprintf("value%d", i%5),
+				"label4", fmt.Sprintf("value%d", i%5),
+				"label5", fmt.Sprintf("value%d", i%5),
+			),
+		})
+	}
+
+	p := NewMemPostings()
+
+	b.ResetTimer()
+	for n := 0; n < b.N; n++ {
+		for i := 0; i < len(series); i++ {
+			p.Add(series[i].id, series[i].lset)
```

Comment:
You are re-adding the same 1000 `id`s N times.
Is this what you intended?  In what situation can this happen in real code?

Reply:
I don't think it's the same `id`, `i` variable from the loop ends up as `id`, doesn't it?
But you're absolutely right that this is not a realistic benchmark. I'm not sure what would be though. I could make the list random, but it would make it harder to compare benchmark runs, plus that might also not be what usually happens.

Reply:
Yes, `i` is the value.  There are 1,000 values, added in order from 0 to 999.

My point is they get added again and again and again to the same `p`.
How could this happen in real life?

Regarding random values, I would have the same question how can this happen.  In Prometheus the ID is generated by incrementing a 64-bit value on every series.

Reply:
Good point, I've updated the benchmark commit, does this make more sense?

Reply:
How can it happen that you need to re-add all the same IDs?

Reply:
I see in https://github.com/prometheus/prometheus/pull/13642#issuecomment-1964001403 you said that a running Prometheus was speeded up.  Maybe you can create a benchmark based on that?

Reply:
How about re-using the benchmark from https://github.com/prometheus/prometheus/pull/13632/files#diff-4f61bbb68838f886c4c6584c06bcc0760eb3176e748ae35988a92ee28d8a0c41 ?

That gives me:

```
goos: linux
goarch: amd64
pkg: github.com/prometheus/prometheus/tsdb/index
cpu: 13th Gen Intel(R) Core(TM) i7-13800H
                   │   main.txt   │               new.txt               │
                   │    sec/op    │   sec/op     vs base                │
MemPostings_Add-20   121.50m ± 3%   85.63m ± 5%  -29.52% (p=0.000 n=20)

                   │   main.txt   │            new.txt             │
                   │     B/op     │     B/op      vs base          │
MemPostings_Add-20   47.80Mi ± 0%   47.80Mi ± 0%  ~ (p=0.547 n=20)

                   │  main.txt   │            new.txt            │
                   │  allocs/op  │  allocs/op   vs base          │
MemPostings_Add-20   306.0k ± 0%   306.0k ± 0%  ~ (p=0.625 n=20)

```

Reply:
That link takes me to tsdb/index/postings.go, with no benchmark visible.
I still have no idea what situation, that can happen in Prometheus, is speeded up by your change.

Reply:
Sorry, wrong link - https://github.com/prometheus/prometheus/pull/13642/files#diff-4f61bbb68838f886c4c6584c06bcc0760eb3176e748ae35988a92ee28d8a0c41

Reply:
Is this the right code?
```
	// Build a list of time series to append for each goroutine.
	for g := 0; g < goroutines; g++ {
		for i := 0; i < seriesCount; i++ {
			// Each time series will have a few base labels.
			builder.Reset(labels.FromStrings(
				"cluster", "dev", // Single common label shared by all time series.
				"job", fmt.Sprintf("job%d", i%10), // A common label with 10 unique values spread over all time series.
				"instance", fmt.Sprintf("instance%d", i%256), // A common label with 256 unique values spread over all time series.
			))
			// And then we will generate a number of random label/value pairs.
			for j := 0; j < labelsCount; j++ {
				builder.Set(fmt.Sprintf("%s%d", randomString(10), j), fmt.Sprintf("%s%d", randomString(20), j))
			}
			series[g][i] = struct {
				id storage.SeriesRef
				ls labels.Labels
			}{
				id: storage.SeriesRef(i),
				ls: builder.Labels(),
			}
		}
	}
```

It's re-using the same set of SeriesRef (`i` from 0 to `seriesCount`) multiple times on different goroutines?
This cannot happen in Prometheus.

Reply:
Can this happen in Prometheus?